routing/payment_lifecycle.go

```diff
@@ -66,158 +67,243 @@ type shardOutcome struct {
 // paymentLifecycle holds all information about the current state of a payment
 // needed to resume if from any point.
 type paymentLifecycle struct {
-	router          *ChannelRouter
-	totalAmount     lnwire.MilliSatoshi
-	paymentHash     lntypes.Hash
-	paySession      PaymentSession
-	timeoutChan     <-chan time.Time
-	currentHeight   int32
-	existingAttempt *channeldb.HTLCAttemptInfo
+	router        *ChannelRouter
+	totalAmount   lnwire.MilliSatoshi
+	paymentHash   lntypes.Hash
+	paySession    PaymentSession
+	timeoutChan   <-chan time.Time
+	currentHeight int32
 }
 
 // resumePayment resumes the paymentLifecycle from the current state.
 func (p *paymentLifecycle) resumePayment() ([32]byte, *route.Route, error) {
 	shardHandler := &shardHandler{
 		router:      p.router,
 		paymentHash: p.paymentHash,
+		quit:        make(chan struct{}),
 	}
 
-	var (
-		lastError error
-		outcome   *shardOutcome
-		attempt   *channeldb.HTLCAttemptInfo
+	// When the payment lifecycle loop exits, we make sure to signal any
+	// sub goroutine to exit, then wait for them to return.
+	defer func() {
+		close(shardHandler.quit)
+		shardHandler.wg.Wait()
+	}()
+
+	// Each time we send a new payment shard, we'll spin up a goroutine
+	// that will collect the result. Either a payment result will be
+	// returned, or a critical error signaling that we should immediately
+	// exit.
+	shardOutcomes := make(chan *shardOutcome)
+	criticalErrs := make(chan error)
+	var lastError error
+
+	// If we had any existing attempts outstanding, we'll start by and
+	// resume spinning up goroutines that'll collect their results and
+	// deliver them to the lifecycle loop below.
+	existingAttempts, err := p.router.cfg.Control.GetAttempts(
+		p.paymentHash,
 	)
+	if err != nil {
+		return [32]byte{}, nil, err
+	}
 
-	// If we have an existing attempt, we'll start by collecting its result.
-	if p.existingAttempt != nil {
-		attempt = p.existingAttempt
+	for _, a := range existingAttempts {
+		attempt := a
 
-		var err error
-		outcome, err = shardHandler.collectResult(attempt)
-		if err != nil {
-			return [32]byte{}, nil, err
-		}
+		shardHandler.wg.Add(1)
+		go func() {
+			defer shardHandler.wg.Done()
+
+			outcome, err := shardHandler.collectResult(
+				&attempt.HTLCAttemptInfo,
+			)
+			if err != nil {
+				select {
+				case criticalErrs <- err:
+				case <-shardHandler.quit:
+				}
+				return
+			}
+
+			select {
+			case shardOutcomes <- outcome:
+			case <-shardHandler.quit:
+			}
+		}()
 	}
 
+	var (
+		success         = false
+		routeFailure    *paymentFailure
+		terminalFailure *paymentFailure
+	)
+
 	// We'll continue until either our payment succeeds, or we encounter a
 	// critical error during path finding.
+Loop:
```

Comment:
style nit: camelCase labels. name could also be more descriptive

Reply:
Not applicable anymore.